cuda_next/include/detail/dimensions.hpp

```diff
@@ -0,0 +1,109 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Next in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA_NEXT_DETAIL_DIMENSIONS
+#define _CUDA_NEXT_DETAIL_DIMENSIONS
+
+#include <cuda/std/mdspan>
+
+namespace cuda_next
+{
+
+template <typename T, size_t... Extents>
+using dimensions = ::cuda::std::extents<T, Extents...>;
+
+// not unsigned because of a bug in ::cuda::std::extents
+using index_type = int;
+
+template <typename T, size_t... Extents>
+struct hierarchy_query_result : public dimensions<T, Extents...>
+{
+  using Dims = dimensions<T, Extents...>;
+  using Dims::Dims;
+  constexpr explicit _CCCL_HOST_DEVICE hierarchy_query_result(const Dims& dims)
+      : Dims(dims)
+  {}
+  static_assert(Dims::rank() > 0);
+
+  T x = Dims::extent(0);
+  T y = Dims::rank() > 1 ? Dims::extent(1) : 1;
+  T z = Dims::rank() > 2 ? Dims::extent(2) : 1;
```

Comment:
I wanted hierarchy_query_result to be sort of a mix of dim3 and extents to make it easier to use for someone used to dim3, for example expressions like hierarchy.index().x or assigning to a dim3. It's probably OK if that is not a constexpr, it can still be accessed as constexpr with extents member functions.
Internally right now all extents objects always have 3 dimensions, because that is what block_dims and other functions creating levels return. I believe even if we end up supporting different number of dimensions we could still make this dim3-like behavior to work, either adding ones for the missing dimensions or collapsing extra ones into z